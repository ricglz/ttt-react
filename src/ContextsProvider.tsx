import React, { useState, useCallback } from 'react';
import { IntlProvider } from 'react-intl';
import { NotificationContainer } from 'react-notifications';
import { BrowserRouter as Router } from 'react-router-dom';
import messages from './messages';

type Locale = string;

function getDirection(locale: Locale) {
  return locale === 'ar' ? ['rtl', 'text-right'] : ['ltr', 'text-left'];
}

type ChildrenComponentProps = {
  locale: Locale;
  changeLocale: (locale: Locale) => void;
};

type Props = {
  ChildrenComponent: React.FC<ChildrenComponentProps>;
  RouterComponent?: any;
};

export default function ContextsProvider({
  ChildrenComponent,
  RouterComponent = Router,
}: Props) {
  const [locale, setLocale] = useState<Locale>('en');

  const changeLocale = useCallback((newLocale) => {
    setLocale(newLocale);
  }, []);

  const [direction, klass] = getDirection(locale);

  return (
    <IntlProvider locale={locale} messages={messages[locale]}>
      <div dir={direction} className={klass}>
        <RouterComponent>
          <ChildrenComponent changeLocale={changeLocale} locale={locale} />
        </RouterComponent>
        <NotificationContainer />
      </div>
    </IntlProvider>
  );
}
