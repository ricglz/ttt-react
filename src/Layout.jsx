import React, { useState, useEffect, useCallback } from 'react';
import { NotificationContainer } from 'react-notifications';
import 'react-notifications/lib/notifications.css';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Contributors from './components/Contributors/Contributors';
import Home from './components/Home/Home';
import Game from './components/Game/Game';
import Login from './components/OnlineGame/Login';
import GameMenu from './components/OnlineGame/GameMenu';
import Tutorial from './components/Tutorial/Tutorial';
import LanguageFooter from './components/Layout/LanguageFooter';
import LanguagePage from './components/Languages/LanguagePage';
import { userPropType } from './constants/props';
import './css/bootstrap.css';
import './css/home.css';
import './css/board.css';
import './css/fonts.css';
import './css/everything.css';
import { getRedirect } from './firebase/firebase';

const renderSinglePlayer = () => <Game ai />;

function updateUser(setUser) {
  getRedirect().then((response) => {
    const tempUser = response.user;
    if (tempUser) {
      const {
        displayName, email, phoneNumber, photoUrl, uid,
      } = tempUser;
      const newUser = {
        name: displayName, email, phoneNumber, photoUrl, uid,
      };
      setUser(newUser);
      localStorage.setItem('user', JSON.stringify(newUser));
    }
  });
}

function renderLayout({
  user, renderGameMenu, renderLanguage, locale,
}) {
  return (
    <>
      <Router>
        <>
          <Switch>
            <Route exact path="/" component={Home} />
            <Route path="/tutorial" component={Tutorial} />
            <Route path="/singleplayer" render={renderSinglePlayer} />
            <Route path="/multiplayer" component={Game} />
            {Object.keys(user).length === 0 ? (
              <Route path="/login" component={Login} />
            ) : (
              <Route
                path="/login"
                render={renderGameMenu}
              />
            )}
            <Route
              path="/online"
              render={renderGameMenu}
            />
            <Route
              path="/language"
              render={renderLanguage}
            />
            <Route path="/contributors" component={Contributors} />
          </Switch>
          <LanguageFooter locale={locale} />
        </>
      </Router>
      <NotificationContainer />
    </>
  );
}

renderLayout.propTypes = {
  user: userPropType.isRequired,
  renderGameMenu: PropTypes.func.isRequired,
  renderLanguage: PropTypes.func.isRequired,
  locale: PropTypes.string.isRequired,
};

function Layout({ locale, changeLocale }) {
  const cachedUser = localStorage.getItem('user');
  const initialUser = cachedUser ? JSON.parse(cachedUser) : {};
  const [user, setUser] = useState(initialUser);

  useEffect(() => {
    updateUser(setUser);
  }, [setUser]);

  const logOut = useCallback(() => {
    setUser({});
    localStorage.removeItem('user');
  }, [setUser]);

  const renderGameMenu = React.useCallback(({ history }) => (
    <GameMenu
      logOut={logOut}
      user={user}
      history={history}
    />
  ), [logOut, user]);
  const renderLanguage = useCallback(({ history }) => (
    <LanguagePage
      locale={locale}
      changeLocale={changeLocale}
      history={history}
    />
  ), [changeLocale, locale]);
  return renderLayout({
    user, renderGameMenu, renderLanguage, locale,
  });
}

Layout.propTypes = {
  changeLocale: PropTypes.func.isRequired,
  locale: PropTypes.string.isRequired,
};

export default Layout;
