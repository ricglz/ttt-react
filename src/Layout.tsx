import React from 'react';
import 'react-notifications/lib/notifications.css';
import LanguageFooter from './components/Layout/LanguageFooter';

import './css/bootstrap.css';
import './css/home.css';
import './css/board.css';
import './css/fonts.css';
import './css/everything.css';

import SwitchWrapper from './SwitchWrapper';

type Props = {
  locale: string;
  changeLocale: (locale: string) => void;
};

function Layout({ locale, changeLocale }: Props) {
  return (
    <>
      <SwitchWrapper changeLocale={changeLocale} currentLocale={locale} />
      <LanguageFooter locale={locale} />
    </>
  );
}

export default Layout;
