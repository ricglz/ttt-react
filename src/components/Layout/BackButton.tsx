import React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router-dom';

type Props = {
  text: string,
  url: string,
};

const BackButton = ({ text, url }: Props) => (
  <div className="col">
    <Link className="btn btn-game btn-lg btn-danger" to={url}>
      <FormattedMessage id={text} defaultMessage="Back" />
    </Link>
  </div>
);

export default BackButton;
