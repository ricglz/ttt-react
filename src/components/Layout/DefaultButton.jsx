import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

const DefaultButton = ({ onClick, text, defaultText }) => (
  <div className="col">
    <button
      type="button"
      className="btn btn-game btn-lg btn-danger"
      onClick={onClick}
    >
      <FormattedMessage id={text} default={defaultText} />
    </button>
  </div>
);

DefaultButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  text: PropTypes.string.isRequired,
  defaultText: PropTypes.string.isRequired,
};

export default DefaultButton;
