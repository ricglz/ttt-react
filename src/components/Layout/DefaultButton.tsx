import React from 'react';
import { FormattedMessage } from 'react-intl';

type Props = {
  onClick: () => void,
  text: string,
  defaultText: string,
};

const DefaultButton = ({ onClick, text, defaultText }: Props) => (
  <div className="col">
    <button
      type="button"
      className="btn btn-game btn-lg btn-danger"
      onClick={onClick}
    >
      <FormattedMessage id={text} defaultMessage={defaultText} />
    </button>
  </div>
);

export default DefaultButton;
