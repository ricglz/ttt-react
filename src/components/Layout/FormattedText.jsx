import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

export const FormattedParagraph = ({ locale }) => (
  <p>
    <FormattedMessage id={locale} />
  </p>
);

FormattedParagraph.propTypes = {
  locale: PropTypes.string.isRequired,
};

export const FormattedHeader = ({ locale }) => (
  <FormattedMessage id={locale}>
    {(txt) => <h1>{txt}</h1>}
  </FormattedMessage>
);

FormattedHeader.propTypes = {
  locale: PropTypes.string.isRequired,
};

export const FormattedHeader2 = ({ locale }) => (
  <h2>
    <FormattedMessage id={locale} />
  </h2>
);

FormattedHeader2.propTypes = {
  locale: PropTypes.string.isRequired,
};
