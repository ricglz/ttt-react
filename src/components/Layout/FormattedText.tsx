import type { Locale } from 'react-i18next';
import { useTranslation } from 'react-i18next';

type SharedProps = {
  locale: Locale;
};

export const FormattedMessage = ({ locale }: SharedProps) => {
  const { t } = useTranslation();
  return <>{t(locale)}</>;
};

export const FormattedParagraph = (props: SharedProps) => (
  <p>
    <FormattedMessage {...props} />
  </p>
);

export const FormattedHeader = (props: SharedProps) => (
  <h1>
    <FormattedMessage {...props} />
  </h1>
);

export const FormattedHeader2 = (props: SharedProps) => (
  <h2>
    <FormattedMessage {...props} />
  </h2>
);
