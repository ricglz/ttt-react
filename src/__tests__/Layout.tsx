import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Layout from '../Layout';
import ContextsProvider from '../ContextsProvider';

jest.mock('../firebase/firebase', () => {
  const originalModule = jest.requireActual('../firebase/firebase');
  return {
    __esModule: true,
    ...originalModule,
    getRedirect: jest.fn(() => null)
  }
})

function Component() {
  return (
    <ContextsProvider
      ChildrenComponent={Layout}
      RouterComponent={MemoryRouter}
    />
  );
}

it('renders without crashing', () => {
  const component = <Component />;
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
