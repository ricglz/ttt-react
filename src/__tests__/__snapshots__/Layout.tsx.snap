// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="text-left"
  dir="ltr"
>
  loading
  <div
    className="row justify-content-center mt-5"
  >
    <div
      className="col-4 col-lg-4 col-xl-4 border-right border-top"
    >
      <a
        className="btn w-100 locale-button"
        href="/language"
        onClick={[Function]}
      >
        en
      </a>
    </div>
  </div>
  <div
    className="notification-container notification-container-empty"
  >
    <div />
  </div>
</div>
`;
